arangod/Pregel/Actor/include/Actor/Actor.h

```diff
@@ -0,0 +1,248 @@
+                                    typename A::State state,
+                                    typename A::Message message) {
+  {std::visit(
+      typename A::template Handler<Runtime>{
+          {pid, pid, std::move(std::make_unique<typename A::State>(state)),
+           runtime}},
+      message)};
+};
+template<typename A>
+concept IsInspectable = requires(typename A::Message message,
+                                 typename A::State state) {
+  {fmt::format("{}", message)};
+  {fmt::format("{}", state)};
+};
+};  // namespace
+template<typename Runtime, typename A>
+concept Actorable = IncludesAllActorRelevantTypes<Runtime, A> &&
+    IncludesName<A> && HandlerInheritsFromBaseHandler<Runtime, A> &&
+    MessageIsVariant<Runtime, A> && IsInspectable<A>;
+
+template<typename Runtime, typename Config>
+requires Actorable<Runtime, Config>
+struct Actor : ActorBase, std::enable_shared_from_this<Actor<Runtime, Config>> {
+  Actor(ActorPID pid, std::shared_ptr<Runtime> runtime,
+        std::unique_ptr<typename Config::State> initialState)
+      : pid(pid), runtime(runtime), state(std::move(initialState)) {}
+
+  Actor(ActorPID pid, std::shared_ptr<Runtime> runtime,
+        std::unique_ptr<typename Config::State> initialState,
+        std::size_t batchSize)
+      : pid(pid),
+        runtime(runtime),
+        state(std::move(initialState)),
+        batchSize(batchSize) {}
+
+  ~Actor() = default;
+
+  auto typeName() -> std::string_view override { return Config::typeName(); };
+
+  void process(ActorPID sender,
+               std::unique_ptr<MessagePayloadBase> msg) override {
+    if (auto* m =
+            dynamic_cast<MessagePayload<typename Config::Message>*>(msg.get());
+        m != nullptr) {
+      push(sender, std::move(m->payload));
+    } else if (auto* n = dynamic_cast<MessagePayload<ActorError>*>(msg.get());
+               n != nullptr) {
+      push(sender, std::move(n->payload));
+    } else {
+      runtime->dispatch(
+          pid, sender,
+          ActorError{UnknownMessage{.sender = sender, .receiver = pid}});
+    }
+  }
+
+  void process(ActorPID sender, velocypack::SharedSlice msg) override {
+    if (auto m =
+            inspection::deserializeWithErrorT<typename Config::Message>(msg);
+        m.ok()) {
+      push(sender, std::move(m.get()));
+    } else if (auto n = inspection::deserializeWithErrorT<ActorError>(msg);
+               n.ok()) {
+      push(sender, std::move(n.get()));
+    } else {
+      auto error =
+          ActorError{UnknownMessage{.sender = sender, .receiver = pid}};
+      auto payload = inspection::serializeWithErrorT(error);
+      if (payload.ok()) {
+        runtime->dispatch(pid, sender, payload.get());
+      } else {
+        fmt::print("Error serializing UnknownMessage");
+        std::abort();
+      }
+    }
+  }
+
+  auto finish() -> void override { finished.store(true); }
+  auto finishedAndIdle() -> bool override {
+    return finished.load() and idle.load();
```

Comment:
Now it's possible it will be never true.

I think it's a bug if you want to support such API

Reply:
Thanks, I fixed that by checking for finished in pushToQueueAndKick instead of work. Like this, when actor is set to finished, no messages can be pushed to queue but the remaining messages in the actor will be worked off.